src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs

```diff
@@ -371,66 +371,113 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
+
+            ArrayTypeSymbol? arrayType = null;
             if (wrappedExpression is BoundArrayCreation ac)
             {
-                var arrayType = (ArrayTypeSymbol)ac.Type;
-                TypeSymbol elementType = arrayType.ElementType.EnumUnderlyingTypeOrSelf();
-
-                // NB: we cannot use this approach for element types larger than one byte
-                //     the issue is that metadata stores blobs in little-endian format
-                //     so anything that is larger than one byte will be incorrect on a big-endian machine
-                //     With additional runtime support it might be possible, but not yet.
-                //     See: https://github.com/dotnet/corefx/issues/26948 for more details
-                if (elementType.SpecialType.SizeInBytes() != 1)
+                // Get the array element type.  This optimization is only supported for core primitive types.
+                arrayType = (ArrayTypeSymbol)ac.Type;
+                specialElementType = arrayType.ElementType.EnumUnderlyingTypeOrSelf().SpecialType;
+                switch (specialElementType)
                 {
-                    return false;
+                    case SpecialType.System_SByte or SpecialType.System_Byte or SpecialType.System_Boolean: // 1 byte
+                        // For primitives that are a single byte in size, a span can point directly to a blob
+                        // containing the constant data.
+                        break;
+
+                    case SpecialType.System_Int16 or SpecialType.System_UInt16 or SpecialType.System_Char: // 2 bytes
+                    case SpecialType.System_Int32 or SpecialType.System_UInt32 or SpecialType.System_Single: // 4 bytes
+                    case SpecialType.System_Int64 or SpecialType.System_UInt64 or SpecialType.System_Double: // 8 bytes
+                        // For primitives that are > 1 byte in size, we can either use CreateSpan if it's available
+                        // or fall back to caching an array.
+                        break;
+
+                    default:
+                        // Not a supported primitive.
+                        return false;
                 }
 
+                // Get the data and number of elements that compose the initialization.
                 elementCount = TryGetRawDataForArrayInit(ac.InitializerOpt, out data);
             }
 
             if (elementCount < 0)
             {
+                // The expression wasn't an array creation, and/or its contents wasn't composed entirely of literals, etc.
+                // and the optimization can't be applied.
                 return false;
             }
 
+            Debug.Assert(arrayType is not null);
+
             if (start is null != length is null)
             {
+                // We can only handle the case where there's a start expression iff there's a length expression.
```

Comment:
Based on the assertion at the top of the method, we cannot enter this condition. Instead of assertion plus this `if` statement, let's just do one `if` statement with `throw new ExceptionUtilities.Unreachable();` at the top of the method.